Cubical/HITs/SetTruncation/Properties.agda

```diff
@@ -20,55 +21,51 @@ private
     ℓ : Level
     A : Type ℓ
 
+rec : {B : Type ℓ} → isSet B → (A → B) → ∥ A ∥₀ → B
+rec Bset f ∣ x ∣₀ = f x
+rec Bset f (squash₀ x y p q i j) =
+  Bset _ _ (cong (rec Bset f) p) (cong (rec Bset f) q) i j
+
 -- lemma 6.9.1 in HoTT book
-elim : {B : ∥ A ∥₀ → Type ℓ} →
-  (Bset : (x : ∥ A ∥₀) → isSet (B x)) →
-  (g : (a : A) → B (∣ a ∣₀)) →
-  (x : ∥ A ∥₀) → B x
+elim : {B : ∥ A ∥₀ → Type ℓ}
+       (Bset : (x : ∥ A ∥₀) → isSet (B x))
+       (g : (a : A) → B (∣ a ∣₀))
+       (x : ∥ A ∥₀) → B x
 elim Bset g ∣ a ∣₀ = g a
-elim {A = A} {B = B} Bset g (squash₀ x y p q i j) =
+elim Bset g (squash₀ x y p q i j) =
   isOfHLevel→isOfHLevelDep 2 Bset _ _
     (cong (elim Bset g) p) (cong (elim Bset g) q) (squash₀ x y p q) i j
 
-setTruncUniversal : {B : Type ℓ} → (isSet B) → (∥ A ∥₀ → B) ≃ (A → B)
-setTruncUniversal Bset = isoToEquiv (iso intro out leftInv rightInv)
+setTruncUniversal : {B : Type ℓ} → isSet B → (∥ A ∥₀ → B) ≃ (A → B)
+setTruncUniversal {B = B} Bset =
+  isoToEquiv (iso (λ h x → h ∣ x ∣₀) (rec Bset) (λ _ → refl) rinv)
   where
-  intro = (λ h a → h ∣ a ∣₀)
-  out = elim (λ x → Bset)
-
-  leftInv : ∀ g → intro (out g) ≡ g
-  leftInv g = refl
-
-  rightInv : ∀ h → out (intro h) ≡ h
-  rightInv h i x =
-    elim (λ x → isProp→isSet (Bset (out (intro h) x) (h x))) (λ a → refl) x i
+  rinv : (f : ∥ A ∥₀ → B) → rec Bset (λ x → f ∣ x ∣₀) ≡ f
+  rinv f i x =
+    elim (λ x → isProp→isSet (Bset (rec Bset (λ x → f ∣ x ∣₀) x) (f x)))
+         (λ _ → refl) x i
 
 elim2 : {B : ∥ A ∥₀ → ∥ A ∥₀ → Type ℓ}
-  (Bset : ((x y : ∥ A ∥₀) → isSet (B x y)))
-  (g : (a b : A) → B ∣ a ∣₀ ∣ b ∣₀)
-  (x y : ∥ A ∥₀) → B x y
-elim2 Bset g =
-  elim (λ _ → isOfHLevelPi 2 (λ _ → Bset _ _))
-    (λ a → elim (λ _ → Bset _ _) (λ b → g a b))
+        (Bset : ((x y : ∥ A ∥₀) → isSet (B x y)))
+        (g : (a b : A) → B ∣ a ∣₀ ∣ b ∣₀)
+        (x y : ∥ A ∥₀) → B x y
+elim2 Bset g = elim (λ _ → isSetΠ (λ _ → Bset _ _))
+                    (λ a → elim (λ _ → Bset _ _) (g a))
 
 elim3 : {B : (x y z : ∥ A ∥₀) → Type ℓ}
-  (Bset : ((x y z : ∥ A ∥₀) → isSet (B x y z)))
-  (g : (a b c : A) → B ∣ a ∣₀ ∣ b ∣₀ ∣ c ∣₀)
-  (x y z : ∥ A ∥₀) → B x y z
-elim3 Bset g =
-  elim2 (λ _ _ → isOfHLevelPi 2 λ _ → Bset _ _ _)
-    (λ a b → elim (λ _ → Bset _ _ _) (λ c → g a b c))
+        (Bset : ((x y z : ∥ A ∥₀) → isSet (B x y z)))
+        (g : (a b c : A) → B ∣ a ∣₀ ∣ b ∣₀ ∣ c ∣₀)
+        (x y z : ∥ A ∥₀) → B x y z
+elim3 Bset g = elim2 (λ _ _ → isSetΠ (λ _ → Bset _ _ _))
+                     (λ a b → elim (λ _ → Bset _ _ _) (g a b))
 
 setTruncIsSet : isSet ∥ A ∥₀
 setTruncIsSet a b p q = squash₀ a b p q
 
-setId : isSet A → ∥ A ∥₀ ≡ A
-setId {A = A} isset =
-  isoToPath (iso (elim {A = A} (λ _ → isset) (λ x → x)) (λ x → ∣ x ∣₀) (λ b → refl) (λ b → idLemma b))
+setTruncId : isSet A → ∥ A ∥₀ ≡ A
+setTruncId {A = A} hS = isoToPath (iso (rec hS (idfun A)) (λ x → ∣ x ∣₀)
+                                       (λ _ → refl) rinv)
```

Comment:
`setId` (like `propId`) could be named better maybe? it makes me think of the `Id` type.

Reply:
I think this Id stands for idempotent. What about setTruncIdem or setTruncIdempotent?

Reply:
Yeah, those would nice!

Reply:
Alright, I'll go for Idempotent